### ISSUE TITLE
panic when sharing rep between threads

### ISSUE THREAD

**Issue by codypiersall** [nanomsg/nng#871]
## NNG & Platform details.

nng head, rev 82afec95, Linux 4.14 ARM 64-bit.  Discovered a panic when bad code is written

## Steps to Reproduce

Generates a panic by sharing a rep0 socket across threads.  I was trying to come up with an example about how things break when you share a rep socket across threads, to demonstrate the usefulness of contexts.  But I was expecting getting some state errors, not panics.

```c
#include <assert.h>
#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

#include <nng/nng.h>
#include <nng/protocol/reqrep0/rep.h>

#define NTHREADS 2

char url[] = "tcp://127.0.0.1:12324";

void * do_recv(void * p_sock) {
    nng_socket sock = *((nng_socket *) p_sock);
    printf("sock id %d\n", nng_socket_id(sock));
    for (;;) {
        void *data;
        size_t size;
        nng_recv(sock, &data, &size, 0);
    }

    return NULL;
}

static pthread_t threads[NTHREADS];

int main() {
    nng_socket sock;
    pthread_attr_t attr;
    assert(nng_rep0_open(&sock) == 0);
    assert(nng_listen(sock, url, NULL, 0) == 0);
    pthread_attr_init(&attr);
    for (int i = 0; i < NTHREADS; i++) {
        pthread_create(&threads[i], &attr, do_recv, (void *) &sock);
    }
    sleep(3);

    return 0;
}
```
Here's what gets printed to the console:

>     panic: appending node already on a list or not inited
>     This message is indicative of a BUG.
>     Report this at https://github.com/nanomsg/nng/issues

I haven't looked in to *why* this is happening but plan to on the next few days if I can.

**Comment by codypiersall** [comment]
(Just made a bunch of edits; this was *not* a segfault, it was a panic.

**Comment by gdamore** [comment]
Thanks, yes, the panic causes abort().  This is clearly a *bug*, as the sockets are supposed to be thread safe (but you should get errors).  Obviously an application that tried to do this is buggy, but nonetheless we should handle this correctly.  I will attempt to diagnose and fix, but this is at lower priority since a well designed application will never encounter it.

**Comment by codypiersall** [comment]
Agreed -- I don't see this as a high priority.  If you don't find time for it I'll take a look in a couple weeks; it seems like a fun code spelunking task.

**Comment by gdamore** [comment]
Actually, this isn't even necessarily a bug in the caller -- because the way you ignore a request is to call cancel again.  It would be highly unusual to do this in *parallel*, but otherwise the test code is not inherently buggy.  I'm researching in NNG...

**Comment by gdamore** [comment]
Well, actually, now that I think about, you can't have two receive operations pending on the same context at the *same* time.  That is utter nonsense.  Fix is coming shortly.

**Comment by gdamore** [comment]
The same bug affects respondent too.  The other protocols look ok though.
